Use `win32-background-launcher` to start the daemon task

anjannath opened the pull request:
Needs:
- [x] https://github.com/crc-org/win32-background-launcher/pull/1
- [x] release `v0.0.0.1` of `win32-background-launcher`

anjannath committed:
msi: add win32-background-launcher to installer

crc-embedder is updated to download win32-background-launcher.exe
from its github release at crc-org/win32-backgroun-launcher so
that it is available while building the msi

it'll be used to start the crc daemon as a background process

anjannath committed:
preflight: use win32-background-launcher to start the daemon task

this removes the powershell script we used earlier to start the daemon
task as a hidden process and uses the background launcher executable
installed by the msi to start the daemon task

updates the preflight checks that installs the daemon task to use the
background launcher executable

anjannath committed:
cache: add win32-background-launcher to the cached binaries

on a released crc version the win32-background-laucher will be installed
in the install location of crc by the msi, but while using a non
installer version this'll help to download the win32-background-launcher
binary by running `crc setup`

this is also helpful for testing different versions of the background
launcher during development

anjannath commented:
/hold 

till required tasks are finished as mentioned in https://github.com/crc-org/crc/pull/3901#issue-1970038820 and https://github.com/crc-org/crc/issues/3879#issue-1951516815

anjannath commented:
/test e2e-microshift-crc

gbraad commented:
can we rename the application on Windows to `crc-background-launcher` instead of the cryptic `win32-...` 
or `background-launcher`. This way people can identify where it belongs to, as paths are not shown in Task Manager.

adrianriobo commented:
Tested on windows 11 https://crcqe-asia.s3.amazonaws.com/nightly/ocp/4.14.3/20231123/11-pro/qe-results/e2e-non-ux.xml cluster behaves as expected

anjannath commented:
> can we rename the application on Windows to `crc-background-launcher` instead of the cryptic `win32-...` or `background-launcher`. This way people can identify where it belongs to, as paths are not shown in Task Manager.

yes, can do, first have to first change the name of the executable in crc-org/win32-background-launcher

edit: created https://github.com/crc-org/win32-background-launcher/pull/3

cfergeau commented:
> can we rename the application on Windows to `crc-background-launcher` instead of the cryptic `win32-...` or `background-launcher`. This way people can identify where it belongs to, as paths are not shown in Task Manager.

Well the launcher is generic and has nothing crc specific. Dunno if there's a way to provide a description to indicate which process it launched?

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -22,18 +22,19 @@ const (
+	DaemonLogFile                 = "crcd.log"
+	CrcLandingPageURL             = "https://console.redhat.com/openshift/create/local" // #nosec G101
+	DefaultAdminHelperURLBase     = "https://github.com/crc-org/admin-helper/releases/download/v%s/%s"
+	CRCWin32BackgroundLauncherURL = "https://github.com/crc-org/win32-background-launcher/releases/download/v%s/%s"
```
I don't think the `CRC` prefix is really useful here, and I'd also drop "win32". This does not need to be exported.

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -47,6 +48,8 @@ const (
 	OpenShiftIngressHTTPPort  = 80
 	OpenShiftIngressHTTPSPort = 443
+
+	Win32BackgroundLauncherExeName = "win32-background-launcher.exe"
```
For admin-helper, we call it `Executable`, not `ExeName`

cfergeau commented on pkg/crc/preflight/preflight_windows.go:
```diff
@@ -100,13 +100,13 @@ var daemonTaskChecks = []Check{
 	{
-		configKeySuffix:    "check-daemon-task-posh-script-present",
-		checkDescription:   "Checking if the daemon task powershell script is present",
-		check:              checkDaemonPoshScript,
```
This was removed in the previous commit, would be nice if each commit builds.

cfergeau reviewed (commented):
overall looks fine, a few comments here and there.

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -187,6 +194,11 @@ func GetKubeAdminPasswordPath() string {
 	return filepath.Join(MachineInstanceDir, DefaultName, "kubeadmin-password")
 }
 
+func GetCRCWindowsBackgroundLauncherDownloadURL() string {
```
GetWin32BackgroundLauncher...?

anjannath committed:
name the executable as crc-background-launcher.exe

the background-launcher upstream is released as
win32-background-launcher.exe since its function
is generic and does nothing crc specific

however when we install it and use it to launch the
daemon process in background on windows it is useful
to name it as crc-background-launcher.exe to make it
easy for users to identify it with crc

cfergeau commented on pkg/crc/cache/cache.go:
```diff
@@ -20,6 +20,7 @@ type Cache struct {
 	executablePath string
 	archiveURL     string
 	version        string
+	nameMismatch   bool
```
`ignoreNameMismatch` would be more explicit.

cfergeau commented on pkg/crc/cache/cache.go:
```diff
@@ -33,8 +34,8 @@ func (e *VersionMismatchError) Error() string {
 
-func newCache(executablePath string, archiveURL string, version string, getVersion func(string) (string, error)) *Cache {
-	return &Cache{executablePath: executablePath, archiveURL: archiveURL, version: version, getVersion: getVersion}
+func newCache(executablePath string, archiveURL string, version string, nameMismatch bool, getVersion func(string) (string, error)) *Cache {
```
Not sure I'd add this arg  by default, might be better to explicitly set it for the background launcher after creating the cache with the default values (`cache.IgnoreNameMismatch()` or `cache.ignoreNameMismatch = true`)

cfergeau commented on pkg/crc/preflight/preflight_daemon_task_check_windows.go:
```diff
@@ -120,7 +86,16 @@ func fixDaemonTaskInstalled() error {
+		return fmt.Errorf("Missing background launcher binary at: %s", constants.Win32BackgroundLauncherPath())
+	}
+
+	binPathWithArgs := fmt.Sprintf(`"%s" daemon`, crcBinPath)
```
This used to be `daemonCmdArgs := `daemon --log-level debug`, is it intentional that you removed the debug logs?

anjannath replied:
yes, we have no way to look at the stdout of the process, and logs to the log file `crcd.log` is always at debug/trace level, so this flag was not useful

cfergeau commented on pkg/crc/cache/cache_windows.go:
```diff
@@ -23,4 +23,6 @@ func NewWin32BackgroundLauncherCache() *Cache {
 			return strings.TrimSpace(stdOut), nil
 		},
 	)
+	cache.ignoreNameMismatch = true
```
Might be worth an explanatory comment? `upstream binary is win32-..., we want to name it crc-...`?

cfergeau reviewed (approved):
2 minor comments, but fine with me with or without changes
/lgtm
/approve
